Use sendWithAsyncReply for NetworkMDNSRegister::RegisterMDNSName

achristensen07 opened the pull request:
#### 7e047990d22bf44da14a4fa24c0ea3f146a3145e
<pre>
Use sendWithAsyncReply for NetworkMDNSRegister::RegisterMDNSName
<a href="https://bugs.webkit.org/show_bug.cgi?id=259657">https://bugs.webkit.org/show_bug.cgi?id=259657</a>
rdar://113156914

Reviewed by Chris Dumez.

This removes an unnecessary message receiver and makes it more clear which code paths need to send a reply.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.cpp:
(WebKit::PendingRegistrationRequest::PendingRegistrationRequest):
(WebKit::registerMDNSNameCallback):
(WebKit::NetworkMDNSRegister::registerMDNSName):
(WebKit::pendingRegistrationRequests): Deleted.
* Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.h:
* Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::didReceiveMessage):
* Source/WebKit/WebProcess/Network/webrtc/WebMDNSRegister.cpp:
(WebKit::WebMDNSRegister::registerMDNSName):
* Source/WebKit/WebProcess/Network/webrtc/WebMDNSRegister.h:
* Source/WebKit/WebProcess/Network/webrtc/WebMDNSRegister.messages.in: Removed.

Canonical link: <a href="https://commits.webkit.org/266461@main">https://commits.webkit.org/266461@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/46059170dd01130ea1a8c03830cafbb43b9b8777

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13874 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14189 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14524 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15612 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13176 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16697 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14271 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15848 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14043 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14653 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11757 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16316 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/13968 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11940 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19552 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13016 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12681 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15892 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13216 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11091 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12488 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16819 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1631 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13059 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

beidson commented on Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.cpp:
```diff
@@ -98,24 +91,23 @@ void NetworkMDNSRegister::closeAndForgetService(DNSServiceRef service)
-        request->connection->connection().send(Messages::WebMDNSRegister::FinishedRegisteringMDNSName { request->requestIdentifier, request->name, WebCore::MDNSRegisterError::DNSSD }, 0);
-
+        request->completionHandler(request->requestIdentifier, request->name, WebCore::MDNSRegisterError::DNSSD);
+        delete request;
```
Instead of 2 `delete request` lines, could you have a single ScopeExit up above handle that?

cdumez replied:
Why can't we have a simple std::unique_ptr local inside this function instead of a ScopeExit? I'd prefer a std::unique_ptr unless there is a good reason not to.

beidson commented on Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.cpp:
```diff
@@ -98,24 +91,23 @@ void NetworkMDNSRegister::closeAndForgetService(DNSServiceRef service)
     }
-    request->connection->connection().send(Messages::WebMDNSRegister::FinishedRegisteringMDNSName { request->requestIdentifier, request->name, { } }, 0);
+    request->completionHandler(request->requestIdentifier, request->name, { });
+    delete request;
```
Instead of 2 `delete request` lines, could you have a single ScopeExit up above handle that?

cdumez replied:
avoidable, as commented above.

beidson commented on Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.cpp:
```diff
@@ -184,12 +173,12 @@ void NetworkMDNSRegister::unregisterMDNSNames(WebCore::ScriptExecutionContextIde
 }
 
-void NetworkMDNSRegister::registerMDNSName(MDNSRegisterIdentifier requestIdentifier, WebCore::ScriptExecutionContextIdentifier documentIdentifier, const String& ipAddress)
+void NetworkMDNSRegister::registerMDNSName(MDNSRegisterIdentifier requestIdentifier, WebCore::ScriptExecutionContextIdentifier documentIdentifier, const String& ipAddress, CompletionHandler<void(MDNSRegisterIdentifier, String, std::optional<MDNSRegisterError>)>&& completionHandler)
```
Maybe I'm wrong, but, wasn't the whole point of MDNSRegisterIdentifier so Networking could communicate back to WebContent when the request was complete?
And now that we're just using a completion handler, we don't need that?

beidson commented on Source/WebKit/WebProcess/Network/webrtc/WebMDNSRegister.cpp:
```diff
@@ -78,10 +78,11 @@ void WebMDNSRegister::registerMDNSName(ScriptExecutionContextIdentifier identifi
     auto& connection = WebProcess::singleton().ensureNetworkProcessConnection().connection();
-    if (connection.send(Messages::NetworkMDNSRegister::RegisterMDNSName { requestIdentifier, identifier, ipAddress }, 0) != IPC::Error::NoError)
-        finishedRegisteringMDNSName(requestIdentifier, { }, MDNSRegisterError::Internal);
+    connection.sendWithAsyncReply(Messages::NetworkMDNSRegister::RegisterMDNSName { requestIdentifier, identifier, ipAddress }, [weakThis = WeakPtr { *this }] (MDNSRegisterIdentifier&& identifier, String&& mdnsName, std::optional<MDNSRegisterError> error) {
```
Does requestIdentifier mean anything anymore, since we have a completion handler?

cdumez commented on Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.cpp:
```diff
@@ -78,17 +79,9 @@ struct PendingRegistrationRequest {
     String name;
     PAL::SessionID sessionID;
     DNSRecordRef record;
+    CompletionHandler<void(MDNSRegisterIdentifier, String, std::optional<WebCore::MDNSRegisterError>)> completionHandler;
```
We probably don't want to pass the String by value.

cdumez commented on Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.cpp:
```diff
@@ -98,24 +91,23 @@ void NetworkMDNSRegister::closeAndForgetService(DNSServiceRef service)
-    auto request = pendingRegistrationRequests().take(reinterpret_cast<uintptr_t>(context));
-    if (!request)
-        return;
+    auto* request = reinterpret_cast<PendingRegistrationRequest*>(context);
```
Seems we could construct a std::unique_ptr for request from the raw pointer here, to avoid manual `delete` calls in several code paths in this function.

cdumez commented on Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.cpp:
```diff
@@ -98,24 +91,23 @@ void NetworkMDNSRegister::closeAndForgetService(DNSServiceRef service)
-        request->connection->connection().send(Messages::WebMDNSRegister::FinishedRegisteringMDNSName { request->requestIdentifier, request->name, WebCore::MDNSRegisterError::DNSSD }, 0);
-
+        request->completionHandler(request->requestIdentifier, request->name, WebCore::MDNSRegisterError::DNSSD);
+        delete request;
```
avoidable, as commented above.

cdumez commented on Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.cpp:
```diff
@@ -98,24 +89,21 @@ void NetworkMDNSRegister::closeAndForgetService(DNSServiceRef service)
-    auto request = pendingRegistrationRequests().take(reinterpret_cast<uintptr_t>(context));
-    if (!request)
-        return;
+    auto request = std::unique_ptr<PendingRegistrationRequest>(reinterpret_cast<PendingRegistrationRequest*>(context));
```
I suspect this would have sufficed:
```
std::unique_ptr request { reinterpret_cast<PendingRegistrationRequest*>(context) };
```

but ok.

achristensen07 replied:
std::unique_ptr doesn't have deduction guides

cdumez replied:
Sad :(

achristensen07 replied:
The deduction guide is omitted because it wouldn't be able to tell the difference between the result of new T and new T[N]

cdumez reviewed (approved):
LGTM

achristensen07 committed:
Use sendWithAsyncReply for NetworkMDNSRegister::RegisterMDNSName
https://bugs.webkit.org/show_bug.cgi?id=259657
rdar://113156914

Reviewed by Chris Dumez.

This removes an unnecessary message receiver and makes it more clear which code paths need to send a reply.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.cpp:
(WebKit::PendingRegistrationRequest::PendingRegistrationRequest):
(WebKit::registerMDNSNameCallback):
(WebKit::NetworkMDNSRegister::registerMDNSName):
(WebKit::pendingRegistrationRequests): Deleted.
* Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.h:
* Source/WebKit/NetworkProcess/webrtc/NetworkMDNSRegister.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::didReceiveMessage):
* Source/WebKit/WebProcess/Network/webrtc/WebMDNSRegister.cpp:
(WebKit::WebMDNSRegister::registerMDNSName):
* Source/WebKit/WebProcess/Network/webrtc/WebMDNSRegister.h:
* Source/WebKit/WebProcess/Network/webrtc/WebMDNSRegister.messages.in: Removed.

Canonical link: https://commits.webkit.org/266461@main
